Sena/doc/create sensor description (#449)
autowarefoundation/autoware-documentation: docs/how-to-guides/integrating-autoware/creating-vehicle-and-sensor-description/creating-sensor-description/creating-sensor-description.md

    ```yaml
    sensor_kit_base_link:
      camera0/camera_link: # Camera
        x: 0.0
        y: 0.0
        z: 0.0
        roll: 0.0
        pitch: 0.0
        yaw: 0.0
      rs_helios_top_base_link: # Lidar
        x: 0.0
        y: 0.0
        z: 0.0
        roll: 0.0
        pitch: 0.0
        yaw: 0.0
      rs_bpearl_front_base_link: # Lidar
        x: 0.0
        y: 0.0
        z: 0.0
        roll: 0.0
        pitch: 0.0
        yaw: 0.0
      gnss_link: # GNSS/INS
        x: 0.0
        y: 0.0
        z: 0.0
        roll: 0.0
        pitch: 0.0
        yaw: 0.0
    ```

As seen above, [x, y, z, roll, pitch, yaw] values are set to 0 for all sensors. This is because the relevant values in the created file will be found after the calibration between the sensor and the `sensor_kit_base_link`. The parent_frame expressed by `sensor_kit_base_link` can belong to any sensor. However, it is preferred that the frame of the sensor located in the center
of the vehicle is parent.

### 2. sensors_calibration.yaml

A file that defines the mounting positions and orientations of sensors with `base_link` as the parent frame.

> The original file is [sensors_calibration.yaml](https://github.com/autowarefoundation/sample_sensor_kit_launch/blob/main/sample_sensor_kit_description/config/sensors_calibration.yaml) included in [sample_sensor_kit_launch](https://github.com/autowarefoundation/sample_sensor_kit_launch/tree/main).
> 
> ??? note "sensors_calibration.yaml for sample_sensor_kit_launch"
> 
>     ```yaml
>     base_link:
>       sensor_kit_base_link:
>         x: 0.9
>         y: 0.0
>         z: 2.0
>         roll: -0.001
>         pitch: 0.015
>         yaw: -0.0364
>         .
>         .
>         .
>     ```
> 

Reviewer: ```suggestion
The default file [sensors_calibration.yaml](https://github.com/autowarefoundation/sample_sensor_kit_launch/blob/main/sample_sensor_kit_description/config/sensors_calibration.yaml) for sample_sensor_kit is given below:

    ```yaml
    base_link:
      sensor_kit_base_link:
        x: 0.9
        y: 0.0
        z: 2.0
        roll: -0.001
        pitch: 0.015
        yaw: -0.0364
        .
        .
        .
    ```

```

In sensors_calibration.yaml file, parent_frame's position relative to base_link is given in euler format as [x, y, z, roll, pitch, yaw].

??? note "sensors_calibration.yaml for tutorial_vehicle_sensor_kit_launch"

    ```yaml
    base_link:
      sensor_kit_base_link:
        x: 0.0
        y: 0.0
        z: 0.0
        roll: 0.0
        pitch: 0.0
        yaw: 0.0
    ```

In this example file, the top_lidar in the center of the vehicle is selected as the parent_frame, and it is prepared in such a way that the position of the parent_frame element relative to the axle of the vehicle is entered.
The [x, y, z, roll, pitch, yaw] values in this file can be entered by base_link - lidar calibration or by measuring from the 3D model.

### 3. sensor_kit.xacro

Resolves the positions of sensors with sensor_kit_base_link as the parent and defines the positions and orientations based on sensor_kit_calibration.yaml in `individual_params`.

> The original file is [sensor_kit.xacro](https://github.com/autowarefoundation/sample_sensor_kit_launch/blob/main/sample_sensor_kit_description/urdf/sensor_kit.xacro)
> included in [sample_sensor_kit_launch](https://github.com/autowarefoundation/sample_sensor_kit_launch/tree/main).

Reviewer: Since sensor_kit.xacro is big file, we can hide by using `note` element.
```suggestion
The defaul [sensor_kit.xacro](https://github.com/autowarefoundation/sample_sensor_kit_launch/blob/main/sample_sensor_kit_description/urdf/sensor_kit.xacro)
 file for sample_sensor_kit is given below:
```

          <parent link="${parent}"/>
          <child link="${sensor_kit_base_link}"/>
        </joint>
        <link name="${sensor_kit_base_link}">
          <origin rpy="0 0 0" xyz="0 0 0"/>
        </link>

        <!-- sensor -->
        <xacro:property name="calibration" value="${xacro.load_yaml('$(arg config_dir)/sensor_kit_calibration.yaml')}"/>

        <!-- lidar -->
        <xacro:VLS-128 parent="sensor_kit_base_link" name="velodyne_top" topic="/points_raw" hz="10" samples="220" gpu="$(arg gpu)">
          <origin
            xyz="${calibration['sensor_kit_base_link']['velodyne_top_base_link']['x']}
                 ${calibration['sensor_kit_base_link']['velodyne_top_base_link']['y']}
                 ${calibration['sensor_kit_base_link']['velodyne_top_base_link']['z']}"
            rpy="${calibration['sensor_kit_base_link']['velodyne_top_base_link']['roll']}
                 ${calibration['sensor_kit_base_link']['velodyne_top_base_link']['pitch']}
                 ${calibration['sensor_kit_base_link']['velodyne_top_base_link']['yaw']}"
          />
          .
          .
          .
      </xacro:macro>
    </robot>
    ```

> The tutorial_vehicle file is [sensor_kit.xacro](https://github.com/leo-drive/tutorial_vehicle_sensor_kit/blob/main/tutorial_vehicle_sensor_kit_description/urdf/sensor_kit.xacro) included in [tutorial_vehicle_sensor_kit](https://github.com/leo-drive/tutorial_vehicle_sensor_kit/tree/main).
> The `sensor_kit.xacro` given below for the tutorial vehicle is prepared for 1 camera,

Reviewer: ```suggestion
The tutorial_vehicle's [sensor_kit.xacro](https://github.com/leo-drive/tutorial_vehicle_sensor_kit/blob/main/tutorial_vehicle_sensor_kit_description/urdf/sensor_kit.xacro) file is  given below for the tutorial vehicle is prepared for 1 camera,
```

          pitch="${calibration['sensor_kit_base_link']['camera0/camera_link']['pitch']}"
          yaw="${calibration['sensor_kit_base_link']['camera0/camera_link']['yaw']}"
          fps="30"
          width="800"
          height="400"
          fov="1.3"
        />

        <!-- gnss -->
        <xacro:imu_macro
          name="gnss"
          parent="sensor_kit_base_link"
          namespace=""
          x="${calibration['sensor_kit_base_link']['gnss_link']['x']}"
          y="${calibration['sensor_kit_base_link']['gnss_link']['y']}"
          z="${calibration['sensor_kit_base_link']['gnss_link']['z']}"
          roll="${calibration['sensor_kit_base_link']['gnss_link']['roll']}"
          pitch="${calibration['sensor_kit_base_link']['gnss_link']['pitch']}"
          yaw="${calibration['sensor_kit_base_link']['gnss_link']['yaw']}"
          fps="100"
        />

      </xacro:macro>
    </robot>
    ```

### 4. sensors.xacro

Resolves the positions of sensors with `base_link` as the parent frame and defines the positions and orientations based on `sensors_calibration.yaml` in `individual_params`.

> The original file is [sensors.xacro](https://github.com/autowarefoundation/sample_sensor_kit_launch/blob/main/sample_sensor_kit_description/urdf/sensors.xacro)
> included in [sample_sensor_kit_launch](https://github.com/autowarefoundation/sample_sensor_kit_launch/tree/main).

Reviewer: ```suggestion
The default [sensors.xacro](https://github.com/autowarefoundation/sample_sensor_kit_launch/blob/main/sample_sensor_kit_description/urdf/sensors.xacro)  file for sample_sensor_kit is given below:
```

??? note "sensors.xacro for sample_sensor_kit_launch"

    ```yaml
    <?xml version="1.0"?>
    <robot name="vehicle" xmlns:xacro="http://ros.org/wiki/xacro">
      <xacro:arg name="config_dir" default="$(find sample_sensor_kit_description)/config"/>
      <xacro:property name="calibration" value="${xacro.load_yaml('$(arg config_dir)/sensors_calibration.yaml')}"/>

      <!-- sensor kit -->
      <xacro:include filename="sensor_kit.xacro"/>
      <xacro:sensor_kit_macro
        parent="base_link"
        x="${calibration['base_link']['sensor_kit_base_link']['x']}"
        y="${calibration['base_link']['sensor_kit_base_link']['y']}"
        z="${calibration['base_link']['sensor_kit_base_link']['z']}"
        roll="${calibration['base_link']['sensor_kit_base_link']['roll']}"
        pitch="${calibration['base_link']['sensor_kit_base_link']['pitch']}"
        yaw="${calibration['base_link']['sensor_kit_base_link']['yaw']}"
      />
      .
      .
      .
    </robot>
    ```

> The tutorial_vehicle file is [sensors.xacro](https://github.com/leo-drive/tutorial_vehicle_sensor_kit/blob/main/tutorial_vehicle_sensor_kit_description/urdf/sensors.xacro) included in [tutorial_vehicle_sensor_kit](https://github.com/leo-drive/tutorial_vehicle_sensor_kit/tree/main) and given below.
> 

Reviewer: ```suggestion
The tutorial_vehicle's [sensors.xacro](https://github.com/leo-drive/tutorial_vehicle_sensor_kit/blob/main/tutorial_vehicle_sensor_kit_description/urdf/sensors.xacro) file is given below.

```

??? note "sensors.xacro for tutorial_vehicle_sensor_kit_launch"

    ```yaml
    <?xml version="1.0"?>
    <robot name="vehicle" xmlns:xacro="http://ros.org/wiki/xacro">
      <xacro:arg name="config_dir" default="$(find sample_sensor_kit_description)/config"/>
      <xacro:property name="calibration" value="${xacro.load_yaml('$(arg config_dir)/sensors_calibration.yaml')}"/>

      <!-- sensor kit -->
      <xacro:include filename="sensor_kit.xacro"/>
      <xacro:sensor_kit_macro
        parent="base_link"
        x="${calibration['base_link']['sensor_kit_base_link']['x']}"
        y="${calibration['base_link']['sensor_kit_base_link']['y']}"
        z="${calibration['base_link']['sensor_kit_base_link']['z']}"
        roll="${calibration['base_link']['sensor_kit_base_link']['roll']}"
        pitch="${calibration['base_link']['sensor_kit_base_link']['pitch']}"
        yaw="${calibration['base_link']['sensor_kit_base_link']['yaw']}"
      />
    </robot>
    ```

## 2. VEHICLE-ID_sensor_kit_launch

VEHICLE-ID_sensor_kit_launch is where the launch files related to sensor startup are stored.

### Folder Structure

> ??? note "VEHICLE-ID_sensor_kit_launch Folder Structure"
> 
>     ```diff
>     VEHICLE-ID_sensor_kit_launch/
>        ├─ config/
>        │     ├─ diagnostic_aggregator/
>        │     │     └─ sensor_kit.param.yaml
>        │     └─ dummy_diag_publisher/
>        │           └─ sensor_kit.param.yaml
>        ├─ data/
>        │     └─ traffic_light_camera.yaml
>        └─ launch/
>              ├─ camera.launch.xml
>              ├─ gnss.launch.xml
>              ├─ imu.launch.xml
>              ├─ lidar.launch.xml
>              ├─ pointcloud_preprocessor.launch.py
>              └─ sensing.launch.xml
>     ```

Reviewer: There is no need for `note` element. Also you can add "+" tag to created wile for `diff`.
```suggestion

    ```diff
    VEHICLE-ID_sensor_kit_launch/
          ├─ config/
          │     ├─ diagnostic_aggregator/
     +   │     │     └─ sensor_kit.param.yaml
          │     └─ dummy_diag_publisher/
     +   │           └─ sensor_kit.param.yaml
          ├─ data/
     +   │     └─ traffic_light_camera.yaml
          └─ launch/
     +        ├─ camera.launch.xml
     +        ├─ gnss.launch.xml
     +        ├─ imu.launch.xml
     +        ├─ lidar.launch.xml
     +        ├─ pointcloud_preprocessor.launch.py
     +        └─ sensing.launch.xml
    ```
```

          path: sensing
          analyzers:
            lidar:
              type: diagnostic_aggregator/AnalyzerGroup
              path: lidar
              analyzers:
                velodyne:
                  type: diagnostic_aggregator/AnalyzerGroup
                  path: velodyne
                  analyzers:
                    health_monitoring:
                      type: diagnostic_aggregator/AnalyzerGroup
                      path: health_monitoring
                      analyzers:
                        connection:
                          type: diagnostic_aggregator/GenericAnalyzer
                          path: connection
                          contains: [": velodyne_connection"]
                          timeout: 3.0

                        temperature:
                          type: diagnostic_aggregator/GenericAnalyzer
                          path: temperature
                          contains: [": velodyne_temperature"]
                          timeout: 3.0

                        rpm:
                          type: diagnostic_aggregator/GenericAnalyzer
                          path: rpm
                          contains: [": velodyne_rpm"]

>                           timeout: 3.0

Reviewer: There is no Velodyne lidar on the tutorial_vehicle, so we can remove this section

            <remap from="~/fix" to="~/nav_sat_fix"/>
            <param from="$(find-pkg-share ublox_gps)/c94_f9p_rover.yaml"/>
          </node>
        </group>

        <!-- Septentrio GNSS Driver -->
        <group if="$(eval &quot;'$(var launch_driver)' and '$(var gnss_receiver)'=='septentrio'&quot;)">
          <include file="$(find-pkg-share septentrio_gnss_driver)/launch/mosaic_x5_rover.launch.xml"/>
        </group>

        <!-- NavSatFix to MGRS Pose -->
        <include file="$(find-pkg-share gnss_poser)/launch/gnss_poser.launch.xml">
          <arg name="input_topic_fix" value="$(var navsatfix_topic_name)"/>
          <arg name="input_topic_orientation" value="$(var orientation_topic_name)"/>

          <arg name="output_topic_gnss_pose" value="pose"/>
          <arg name="output_topic_gnss_pose_cov" value="pose_with_covariance"/>
          <arg name="output_topic_gnss_fixed" value="fixed"/>

          <arg name="use_gnss_ins_orientation" value="true"/>

          <arg name="gnss_frame" value="gnss_link"/>
        </include>
      </group>
    </launch>

>     ```

Author: I think we should explain all launch files at the same time.

          <arg name="output_topic_gnss_fixed" value="fixed"/>

          <arg name="base_frame" value="base_link"/>
          <arg name="gnss_base_frame" value="gnss_ins_base_link"/>
          <arg name="gnss_frame" value="GNSS_INS/gnss_ins_link"/>
          <arg name="map_frame" value="map"/>

          <arg name="coordinate_system" value="$(var coordinate_system)"/>
          <arg name="use_gnss_ins_orientation" value="true"/>
          <arg name="height_system" value="$(var height_system)"/>

          <arg name="buff_epoch" value="4"/>
          <arg name="gnss_pose_pub_method" value="$(var gnss_pose_pub_method)"/>
        </include>
      </group>

      <group>
        <push-ros-namespace namespace="imu"/>
        <include file="$(find-pkg-share imu_corrector)/launch/imu_corrector.launch.xml">
          <arg name="input_topic" value="/sensing/gnss/clap/ros/imu"/>
          <arg name="output_topic" value="imu_data"/>
          <arg name="param_file" value="$(find-pkg-share individual_params)/config/$(var vehicle_id)/tutorial_vehicle_sensor_kit/imu_corrector.param.yaml"/>
        </include>
      </group>
    </launch>
    ```

### 6. imu.launch.xml

> imu.launch.xml is the launch file related to starting the IMU driver.
> The original file is [imu.launch.xml](https://github.com/autowarefoundation/sample_sensor_kit_launch/blob/main/sample_sensor_kit_launch/launch/imu.launch.xml)
> included in [sample_sensor_kit_launch](https://github.com/autowarefoundation/sample_sensor_kit_launch/tree/main) and given below.

Reviewer: ```suggestion
imu.launch.xml is the launch file related to starting the IMU driver.
The default  [imu.launch.xml](https://github.com/autowarefoundation/sample_sensor_kit_launch/blob/main/sample_sensor_kit_launch/launch/imu.launch.xml) file for sample_sensor_kit is given below.
```
